New functionalities to display the state date when invoking the fuzzy loader function

ChrisGVE opened the pull request:
## Functionality
The new functionality primarily adds the state's last modification date when showing the list of states in the fuzzy loader function. It is configured via the fuzzy loader's `opts` table, which is given sensible defaults.

In addition, the modification also manages long state names that may be longer than the available screen size, whether the date is included or not.

## Compatibility
The new feature has been tested for macOS, Windows 11, and Linux. It relies only on standard commands of these operating systems.

## Documentation
`README.md` has been modified to document this new feature.

This pull request relates to #82

MLFlexer commented on plugin/resurrect/fuzzy_loader.lua:
```diff
@@ -57,32 +77,375 @@ function pub.fuzzy_load(window, pane, callback, opts)
+			local oversize = math.max(0, width - win_width)
+
+			if oversize == 0 then
+				-- No oversize for this line, thus we keep it as is (though it shouldn't occur when we use this function)
```
Is this an unreachable state? Or why shouldn't it occur?

ChrisGVE replied:
Because I use an estimate to trigger this formatting this could happen

MLFlexer commented on plugin/resurrect/fuzzy_loader.lua:
```diff
@@ -57,32 +77,375 @@ function pub.fuzzy_load(window, pane, callback, opts)
+			-- check the overall width against the available width
+			local width = label.name_len + label.date_len + 4
+			-- `oversize` is the number of character we should remove
+			local oversize = math.max(0, width - win_width)
```
I'm not too keen on this name, what about something like `excess_chars`? or `excess_width`?

ChrisGVE replied:
I've changed the name to `overflow_chars`

MLFlexer commented on plugin/resurrect/fuzzy_loader.lua:
```diff
@@ -57,32 +77,375 @@ function pub.fuzzy_load(window, pane, callback, opts)
+				-- No oversize for this line, thus we keep it as is (though it shouldn't occur when we use this function)
+				return label.name_fmt .. label.date_fmt
+			else
+				-- we need to save a `oversize` character; first check if the padding can be used
```
This is a bit hard to understand, maybe something like this comment instead
```suggestion
				-- we need to remove characters; first check if the padding can be used
```

ChrisGVE commented on plugin/resurrect/utils.lua:
```diff
@@ -0,0 +1,54 @@
+-- getting screen dimensions
+---@return number
+function utils.get_current_window_width()
+	local windows = wezterm.mux.all_windows()
```
Thanks for the tip!

ChrisGVE commented on plugin/resurrect/fuzzy_loader.lua:
```diff
@@ -57,32 +77,375 @@ function pub.fuzzy_load(window, pane, callback, opts)
+	-- Optimized recursive JSON file finder for all platforms
+	---@param base_path string starting path from which the recursive search takes place
+	---@return string
+	local function find_json_files_recursive(base_path)
```
Well, I've been helped by some AI 🙂
Unlike Linux or macOS, which have silent executions, this wasn't the case with Windows. I always had some transient window frames while the command was executed. The execution was short enough not to require the terminal/powershell window to be fully built, but sufficiently for its frame to appear, which was ugly.
This is where the VBS came up, and now it works without these ugly artefacts.
I left the temporary files as suggested by the AI and did not consider whether they could be avoided altogether 😀.

ChrisGVE commented:
Thanks for the thorough review. Indeed, `fuzzy_load` has some room for refactoring, but I followed the same path since you originally wrote a self-sufficient function with its helpers in its scope. Stylistically, it could benefit from some clean-up, but I wanted to submit the PR as is. Later, we can always go back and improve.

ChrisGVE commented on plugin/resurrect/fuzzy_loader.lua:
```diff
@@ -57,32 +77,375 @@ function pub.fuzzy_load(window, pane, callback, opts)
+		local cmd
+
+		if utils.is_windows then
+			-- For Windows, use VBS for better performance and truly invisible execution
```
Good questions. I mentioned in another response that transient window frames appeared when spinning cmd or PowerShell. But that wasn't all; by running them, I encountered quite a bit of overhead, making the function simply not practical (note that I run this on a virtual Win 11 without WSL), so waiting for more than a second to get the list of states was not acceptable. And here came the AI rewrite (I think I used Deepseek), which came a bit from the left field but worked effectively with no side effects.

ChrisGVE commented on plugin/resurrect/fuzzy_loader.lua:
```diff
@@ -57,32 +77,375 @@ function pub.fuzzy_load(window, pane, callback, opts)
+			local handle = io.open(temp_vbs, "w")
+			if handle == nil then
+				wezterm.emit("resurrect.error", "Could not create temporary Windows process")
+				return ""
```
I've created three helper functions to write/read and execute commands in `utils.lua` such that they can be reused, I've taken the opportunity to change your save function in `state_manager.lua`. It also reduced the length of `fuzzy_loader.lua`

MLFlexer commented on plugin/resurrect/utils.lua:
```diff
@@ -0,0 +1,124 @@
+utils.is_mac = (wezterm.target_triple == "x86_64-apple-darwin" or wezterm.target_triple == "aarch64-apple-darwin")
+utils.separator = utils.is_windows and "\\" or "/"
+
+local constants = require("utils.constants")
```
Not used?
```suggestion
```

MLFlexer commented on plugin/resurrect/utils.lua:
```diff
@@ -0,0 +1,124 @@
+
+local constants = require("utils.constants")
+
+-- Helper function to remove formatting sequence in strings
+---@param str string
+---@return string
+function utils.strip_format(str)
+	local clean_str, _ = str:gsub(string.char(27) .. "%[[^m]*m", "")
+	return clean_str
+end
```
Is this only for date strings? If so, then i think a better name would be `strip_data_format`, as others might use it for non-date strings in the future

MLFlexer commented on plugin/resurrect/utils.lua:
```diff
@@ -0,0 +1,124 @@
+	if suc then
+		return stdout, true
+	else
+		return "", suc, err
```
One to many returns
```suggestion
		return "", err
```

ChrisGVE replied:
You are absolutely right here; I did a poor job with this function and the others. I'll correct them and check the rest of the code base to see where it can benefit from refactoring these functions.

MLFlexer commented on plugin/resurrect/utils.lua:
```diff
@@ -0,0 +1,124 @@
+	if suc then
+		return stdout, true
+	else
+		return "", suc, err
```
Same as above
```suggestion
		return "", err
```

MLFlexer commented on plugin/resurrect/utils.lua:
```diff
@@ -0,0 +1,124 @@
+---@param path string
+function utils.ensure_folder_exists(path)
+	if utils.is_windows then
+		os.execute('mkdir /p "' .. path:gsub("/", "\\" .. '"'))
+	else
+		os.execute('mkdir -p "' .. path .. '"')
```
Better to use wezterm.background_child_process for executing processes, as they have invisible execution
```suggestion
		wezterm.background_child_process({"mkdir", "/p", "'" .. path:gsub("/", "\\" .. '"') .. "'"})
	else
		wezterm.background_child_process({"mkdir", "-p", "'" .. path .. "'"})
```

MLFlexer replied:
Ignore this for now, this can be done in a later iteration of the code

MLFlexer commented on plugin/resurrect/fuzzy_loader.lua:
```diff
@@ -15,35 +27,391 @@ function pub.get_default_fuzzy_load_opts()
+-- Optimized recursive JSON file finder for all platforms
+---@param base_path string starting path from which the recursive search takes place
+---@return string|nil
+local function find_json_files_recursive(base_path)
```
We might be able to execute the vb script in a [run_child_process](https://wezterm.org/config/lua/wezterm/run_child_process.html) as they have invisible execution as well, but I think improvements to this can be done in another PR, as this should be merged soon as it is already big enough for a single PR :smile:

ChrisGVE replied:
I agree, I have written in utils a function to execute via `popen` and return either stdout or err, I don't recall if I made use of it or if I just complemented the other utility functions, in any case, I agree that executing stuff would need a review, but later 😉

ChrisGVE replied:
Ok I realized that I wasn't respecting the separation of concerns, so I moved `read_file` and `write_file` to `file_io.lua`. Now looking at your `load_json`, I see that you are loading the file line by line in a table but then you are concatenating the table? Is there a reason you don't just grab the file in one go instead?

MLFlexer commented on plugin/resurrect/fuzzy_loader.lua:
```diff
@@ -15,35 +27,391 @@ function pub.get_default_fuzzy_load_opts()
+	end
+
+	-- Execute the command and capture stdout for non-Windows
+	stdout, suc, err = utils.execute(cmd)
+
+	if suc then
+		return stdout
```
This needs to match the execute returns
```suggestion
	stdout, err = utils.execute(cmd)

	if err then
		return stdout
```

MLFlexer commented on plugin/resurrect/fuzzy_loader.lua:
```diff
@@ -15,35 +27,391 @@ function pub.get_default_fuzzy_load_opts()
+		-- Execute using launcher (completely hidden)
+		os.execute("wscript.exe //nologo " .. launcher_vbs)
+
+		stdout, suc, err = utils.read_file(temp_out)
+
+		-- Clean up temp files
+		os.remove(temp_vbs)
+		os.remove(launcher_vbs)
+		os.remove(temp_out)
+
+		if suc then
+			return stdout
```
This needs to match the execute returns
```suggestion
		stdout, err = utils.read_file(temp_out)

		-- Clean up temp files
		os.remove(temp_vbs)
		os.remove(launcher_vbs)
		os.remove(temp_out)

		if err == nil then
			return stdout
```

MLFlexer commented on plugin/resurrect/fuzzy_loader.lua:
```diff
@@ -15,35 +27,391 @@ function pub.get_default_fuzzy_load_opts()
+	local overflow_chars = math.max(0, width - win_width)
+
+	if overflow_chars == 0 then
+		-- No oversize for this line, thus we keep it as is
```
```suggestion
		-- No overflow_chars for this line, thus we keep it as is
```

MLFlexer commented on plugin/resurrect/fuzzy_loader.lua:
```diff
@@ -15,35 +27,391 @@ function pub.get_default_fuzzy_load_opts()
+		-- we need to remove characters; first check if the padding can be used
+		if label.padding_len ~= 0 then
+			local new_len = math.max(0, label.padding_len - overflow_chars)
+			overflow_chars = overflow_chars - (label.padding_len - new_len) -- update the oversize
```
```suggestion
			overflow_chars = overflow_chars - (label.padding_len - new_len) -- update the overflow_chars
```

MLFlexer commented on plugin/resurrect/fuzzy_loader.lua:
```diff
@@ -15,35 +27,391 @@ function pub.get_default_fuzzy_load_opts()
+		-- we tackle the filename reducing it to a length with a minimum of `min_filename_len`
+		if overflow_chars ~= 0 then
+			-- new we need to apply the size reduction to the filename, our strategy:
+			-- remove the `oversize` from the middle of the filename string and
```
```suggestion
			-- remove the `overflow_chars` from the middle of the filename string and
```

MLFlexer commented on plugin/resurrect/fuzzy_loader.lua:
```diff
@@ -15,35 +27,391 @@ function pub.get_default_fuzzy_load_opts()
+			overflow_chars = overflow_chars - reduction
+			label.filename_raw = utils.replace_center(label.filename_raw, reduction, str_pad)
+		end
+		-- do we still have an oversize? we can do something only if we have a date, otherwise we did our best
```
```suggestion
		-- do we still have an overflow_chars? we can do something only if we have a date, otherwise we did our best
```

MLFlexer commented on plugin/resurrect/fuzzy_loader.lua:
```diff
@@ -15,35 +27,391 @@ function pub.get_default_fuzzy_load_opts()
+	-- Parse the stdout and construct the file table
+	for line in stdout:gmatch("[^\n]+") do
+		local epoch, type, file = line:match("%s*(%d+)%s+.+[/\\]([^/\\]+)[/\\]([^/\\]+%.json)$")
+		if epoch and file and type and type == type then
```
unnecessary check on itself
```suggestion
		if epoch and file and type then
```

MLFlexer commented on plugin/resurrect/fuzzy_loader.lua:
```diff
@@ -15,35 +27,391 @@ function pub.get_default_fuzzy_load_opts()
+
+	-- Parse the stdout and construct the file table
+	for line in stdout:gmatch("[^\n]+") do
+		local epoch, type, file = line:match("%s*(%d+)%s+.+[/\\]([^/\\]+)[/\\]([^/\\]+%.json)$")
```
What is epoch, the date in number format? A comment to clarify this might be useful, as it isn't intuitive what this represents :smile:

ChrisGVE replied:
I've commented it. The advantage of using the epoch over a formatted date is that you only deal with a number that counts the time since a fixed date (OS defined), and then when formatting it for the host OS you get whatever format you want.

MLFlexer commented on plugin/resurrect/fuzzy_loader.lua:
```diff
@@ -57,32 +425,22 @@ function pub.fuzzy_load(window, pane, callback, opts)
-	end
+	local stdout = find_json_files_recursive(folder)
 
-	if not opts.ignore_workspaces then
```
ignore_workspaces/windows/tabs is not used as far as i can tell.
Shouldn't they be added to the `find_json_files_recursive` or the `insert_choices`?

ChrisGVE replied:
It is used. I've fought quite a bit the last two days, but I'm happy now because all work as intended. I'll give you the run down in the overall message

MLFlexer commented on plugin/init.lua:
```diff
@@ -1,48 +1,45 @@
-	local path2 = "httpssCssZssZsgithubsDscomsZsMLFlexersZsresurrectsDsweztermsZs"
-	return directory_exists(path2) and path2 or path1
+	local paths = {
+		-- chrisgve
+		"httpssCssZssZsgithubsDscomsZschrisgvesZsresurrectsDswezterm",
+		"httpssCssZssZsgithubsDscomsZschrisgvesZsresurrectsDsweztermsZs",
+		"httpsCssZssZsgithubsDscomsZschrisgvesZsresurrectsDswezterm",
+		"httpsCssZssZsgithubsDscomsZschrisgvesZsresurrectsDsweztermsZs",
```
```suggestion
```

ChrisGVE replied:
I'll keep this on until we merge in case we still have changes to make

ChrisGVE commented:
HelloI’ve worked on most issues but I’m revisiting once more the formatting of the state list to make it more efficient. I’ve simplified and removed loops and added some simple optimizations so that the handling of file name does not waste unnecessary cycles. It starts to work well but I’m still debugging the last leg of formatting. Hopefully I’m done tomorrow (well later today) and you can have another look.CheersSent from my iPhoneOn 30 Mar 2025, at 11:31, Malthe Larsen ***@***.***> wrote:﻿
@MLFlexer requested changes on this pull request.

There are some problems with the return types which dont match the type annotations and/or the expected return
And then the ignore options do not get used.
Using the utils module more when doing io has greatly improved the readability I think, so great job!
Again, thanks for working on this! 😄  I think it is close to being merged, we just need the last few things with the options and the returns 🚀

—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you authored the thread.Message ID: ***@***.***>

ChrisGVE commented:
So again, quite a big review since I've made numerous changes to the code.
Besides the changes you noted, I've reviewed the logic several times and improved its performance to a good level. Let me outline the process and my thinking:

1. The data collection is done by `find_json_files_recursive,` which only depends on the host OS. Since these commands should run fast, there is no filtering beyond fetching only JSON files.
2. The data treatment and formatting are now fully contained within `insert_choices`:

  - First, we loop and parse every line of stdout that we have gotten before. In this loop, we collect the epoch (I've made a comment), the type, and the file name. Types that are ignored are rejected at this point. The parsing takes the opportunity to calculate some formatting metrics that will be used later (this is done once only) and stores the data into a `files` dictionary by type.
  - Second, we have a double loop, first on types and second on files, that will gather all the data, format the choices, and provide the IDs for the InputSelector. Note that the first loop creates an arrangement where the types are ordered by workspace, window, and then tab. Second, the next loop won't run if a type is ignored since the dictionary is empty.
4. Options
  - As mentioned above, we have the option to select which type is included
  - There are options to include, format the date, and format its appearance in the selection
  - There is an option to ignore or not the window width, if it is set to true, it will simply align the dates to the longest file name regardless whether the screen is too narrow or not, if it is set to false it will always use the full width adaptatively: if there is enough room the dates will be further away (but the dots make it easy to see which is which) and if there is not enough room the separation between filenames and dates will shrink to the point where if necessary the name will be also reduced by removing some character in the middle of it and replacing them with a pad, the reduction is capped to a minimum length. Then the names are no longer cut (thus we'll see misalignments on the screen when the dates are displayed, but as I tested it, I don't expect it to arrive often even with an old 80-column monitor😂). The size management will be done, whether the dates are displayed or not, only acting on the names in that scenario.
5. Utils
I brought in two useful functions: `deepcopy` and `tbl_deep_extend`. The first is self-explanatory, the second comes from the neovim API. It allows managing default and user options, generally by prioritizing user options over the default ones, and you can think about using it elsewhere if that's useful to you.
I removed `read_file` and `write_file` from utils to file_io for a better separation of concerns. I've reviewed some of the code, but I've not deduplicated everything. This is a low priority.

I think I'm done for this PR. It's been a journey, but I've learn a lot. One thing I did learn is that you are an outstanding reviewer, and I hope I can get as good at my code. Thanks a bunch!

MLFlexer commented on plugin/resurrect/fuzzy_loader.lua:
```diff
@@ -287,119 +182,130 @@ local function insert_choices(stdout, opts)
+	else
+		-- During the selection view, InputSelector will take 4 characters on the left and 2 characters
+		-- on the right of the window
+		used_width = utils.get_current_window_width() - 6
```
It seems as this is more available width instead of used? so maybe `available_width` is a better name?

This is very much nitpicking, so if you disagree then just resolve this

MLFlexer commented on plugin/resurrect/fuzzy_loader.lua:
```diff
@@ -457,6 +350,7 @@ function pub.fuzzy_load(window, pane, callback, opts)
 		}),
 		pane
 	)
+	fmt_cost = {} -- we need to reinitialize this since next call might be with different options, including formatting
```
If this is reset every time, then maybe it should be moved into the insert function? Instead of being global variable

MLFlexer reviewed (changes requested):
It was much faster to review this time, which should be a testament to the code being in a really good state! :rocket: 

I have some minor comments, which could be fixed with an lsp rename and moving a variable :smile: 

Other than that, it looks really good now, thanks a bunch for your work!

Will merge if you resolve the comments :smile:

ChrisGVE commented:
Thanks for your words of encouragement. Your comments have been spot on, and I've made the changes as requested. I think we are good to merge, I leave it to you to remove the unnecessary lines in `init.lua` such that until you merge, my version is still working :)

MLFlexer commented:
It's merged now, thanks again! :smile:
